Featured Product: Allow to resize the block height

ryelle opened the pull request:
Fixes #319 – This adds [the `ResizableBox` component](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/resizable-box) around the featured product block, so that the block can be vertically resized by clicking & dragging a handle at the bottom of the block. This uses a min-height of 500px (the current min height), and has no upper limit.

Changing out the background image for a custom image will keep the customized height.

#### Accessibility

The resize handle doesn't appear to be keyboard-friendly, a problem in Gutenberg too: https://github.com/WordPress/gutenberg/issues/10595

### Screenshots

<img width="598" alt="screen shot 2019-02-04 at 2 12 51 pm" src="https://user-images.githubusercontent.com/541093/52231097-0a4c9d80-2887-11e9-81d0-181f29714cb4.png">

### How to test the changes in this Pull Request:

1. Add a Featured Product block
2. Pick a product with or without an image
3. Expect: You can grab the handle and make the block larger
4. Expect: Once larger, you can also grab the handle to make the block smaller again (but it will stop at 500px tall)
5. Leave the block taller than 500px
6. Change the image for a custom image
7. Expect: the block stays the same height
8. Expect: you can still drag to change the height
9. Save the post, view it
10. Expect: the block is using the correct, customized height

ryelle committed:
Add ResizeableBox wrapper to component to adjust height

ryelle committed:
Use the fixed height on the front end display of the block

ryelle committed:
Fix display of resize handle

coderkevin commented on assets/js/blocks/featured-product/block.js:
```diff
@@ -39,6 +40,11 @@ import {
+/**
+ * The min-height for the block content.
+ */
+const MIN_HEIGHT = 500;
```
Should this be in the block settings? Seems like this could be theme-sensitive.

coderkevin commented on assets/js/blocks/featured-product/index.js:
```diff
@@ -50,6 +50,14 @@ registerBlockType( 'woocommerce/featured-product', {
+		 */
+		height: {
+			type: 'number',
+			default: 500,
```
It'd be nice if this used the same variable as above so it's only in one place.

ryelle replied:
I agree these should be extendable, but technically these don't have to be related – MIN_HEIGHT could be a smaller value if we want to allow folks to shrink down the block.

coderkevin commented on assets/js/blocks/featured-product/style.scss:
```diff
@@ -13,11 +13,19 @@
 	}
 
+	.components-resizable-box__handle {
+		z-index: 10;
```
I'm a bit surprised we don't use sass variables for z-index settings. 🤔

ryelle replied:
We could add them, if you want– it didn't seem a priority at the start. Want to add an issue about it?

coderkevin commented on includes/blocks/class-wc-block-featured-product.php:
```diff
@@ -29,6 +29,7 @@ class WC_Block_Featured_Product {
 		'align'        => 'none',
 		'contentAlign' => 'center',
 		'dimRatio'     => 50,
+		'height'       => 500,
```
Would be awesome if the 500 was set here in the PHP, then passed through the block settings and used from there. Then it could be hooked and changed if needed.

coderkevin reviewed (approved):
Code looks good other than the suggestion to include the 500 px figure in the block settings or in some way that can be hooked by a theme. Pre-approving.

coderkevin commented:
Forgot to mention, tests out fine too. Good job!

jameskoster reviewed (approved):
The handle works as expected. 

I have some feedback on the CSS though. Not sure if that is coming from the core component and whether it should be addressed here or not, but here goes:

I noticed it sets a `height` value, both for the default height and the custom height. This is problematic in two situations;

1. Products with long descriptions - [content can be hidden](https://cldup.com/mnMTGp4oKt-3000x3000.png).
2. Mobile - as content narrows on mobile is generally gets taller too. The height set for desktop may not also suit mobile and lead to hidden content.

The simplest solution would _seem_ to be to use `min-height` instead of `height`.

ryelle committed:
Allow the default & min featured product heights to be customized by setting a new theme support keyword, `featured_block`.

ryelle committed:
Use `min-height` on the front end display, so content is not clipped on smaller screens

ryelle committed:
Add a wrapper div to the content, to prevent it from breaking out of the block display

ryelle commented:
I've added support for new theme settings, `featured_block`, which can set default and minimum block heights (thinking we'll use the same settings for Featured Product and Featured Product Category) - this can be set in a theme like this:

```php
function example_store_setup_theme() {
	add_theme_support( 'woocommerce', array(
		'featured_block' => array(
			'min_height' => 300,
			'default_height' => 400,
		),
	) );
}
add_action( 'after_setup_theme', 'example_store_setup_theme' );
```

I've also switched the front end display to use min-height, so that on smaller screens the content won't be cut off. In the editor, you can still appear to chop off content, but on the front end it'll expand to the full height of the content. While this isn't exactly a great live preview, we can't change the `ResizeableBox` component from using `height` directly - dragging the height stops working, if I try.

@jameskoster @coderkevin  Can you both take another look at this?

jameskoster commented:
Shame about the preview, but that's a core issue. Frontend works great :)

coderkevin reviewed (approved):
The code looks good to me. Thanks!
